packages/common/src/store/buy-usdc/sagas.ts

```diff
@@ -0,0 +1,248 @@
+
+  // Wait for on ramp finish
+  const result = yield* race({
+    success: take(onRampSucceeded),
+    canceled: take(onRampCanceled)
+  })
+
+  // If the user didn't complete the on ramp flow, return early
+  if (result.canceled) {
+    yield* call(
+      track,
+      make({ eventName: Name.BUY_USDC_ON_RAMP_CANCELED, provider })
+    )
+    return {}
+  }
+  yield* call(
+    track,
+    make({ eventName: Name.BUY_USDC_ON_RAMP_SUCCESS, provider })
+  )
+
+  // Wait for the SOL funds to come through
+  const newBalance = yield* call(pollForBalanceChange, audiusBackendInstance, {
+    tokenAccount,
+    initialBalance,
+    retryDelayMs,
+    maxRetryCount
+  })
+
+  // Check that we got the requested SOL
+  const purchasedAmount = new BN(newBalance).sub(new BN(initialBalance))
+  if (purchasedAmount !== new BN(desiredAmount.amount)) {
+    console.warn(
+      `Warning: Purchase USDC amount differs from expected. Actual: ${new BN(
+        newBalance
+      )
+        .sub(new BN(initialBalance))
+        .toNumber()} Wei. Expected: ${desiredAmount.uiAmountString} USDC.`
+    )
+  }
+
+  return { newBalance }
+}
+
+function* doBuyUSDC({
+  payload: { desiredAmount }
+}: ReturnType<typeof onRampOpened>) {
+  const audiusBackendInstance = yield* getContext('audiusBackendInstance')
+  const reportToSentry = yield* getContext('reportToSentry')
+  const userRootWallet = yield* call(
+    getCurrentUserWallet,
+    audiusBackendInstance
+  )
+  const { track, make } = yield* getContext('analytics')
+  const provider = yield* select(getBuyUSDCProvider)
+  try {
+    if (provider !== OnRampProvider.STRIPE) {
+      throw new Error('USDC Purchase is only supported via Stripe')
+    }
+
+    yield* put(setVisibility({ modal: 'StripeOnRamp', visible: true }))
+
+    // Record start
+    yield* call(
+      track,
+      make({ eventName: Name.BUY_USDC_ON_RAMP_OPENED, provider })
+    )
+
+    // Setup
+
+    // Get config
+    const { retryDelayMs, maxRetryCount } = yield* call(getBuyUSDCRemoteConfig)
+
+    const userBank = yield* getUSDCUserBank()
+
+    // STEP ONE: Wait for purchase
+    // Have to do some typescript finangling here due to the "race" effect in purchaseStep
+    // See https://github.com/agiledigital/typed-redux-saga/issues/43
+    const { newBalance } = (yield* call(purchaseStep, {
+      provider,
+      desiredAmount,
+      tokenAccount: userBank,
+      retryDelayMs,
+      maxRetryCount
+    }) as unknown as ReturnType<typeof purchaseStep>)!
+
+    // If the user canceled the purchase, stop the flow
+    if (newBalance === undefined) {
+      return
+    }
+
+    yield* put(buyUSDCFlowSucceeded())
+
+    // Record success
+    yield* call(
+      track,
+      make({
+        eventName: Name.BUY_USDC_SUCCESS,
+        provider,
+        requestedAmount: desiredAmount.uiAmount
+        // TODO: actualAmount
+      })
+    )
+  } catch (e) {
+    const stage = yield* select(getBuyUSDCFlowStage)
```

Comment:
We don't really have too many "stages" for BuyUSDC, so probably ok to cut stage stuff to reduce complexity if you want